lxd/metrics/api_rates.go

```diff
@@ -0,0 +1,93 @@
+package metrics
+
+import (
+	"net/url"
+	"sync/atomic"
+
+	"github.com/canonical/lxd/shared/entity"
+)
+
+// RequestStatus represents the a completed request status category
+type requestStatus string
+
+const (
+	serverError requestStatus = "server_error"
+	clientError requestStatus = "client_error"
+	success     requestStatus = "succeeded"
+)
+
+// PossibleRequestStatus is a slice that includes all the possible request status categories
+var PossibleRequestStatus = []requestStatus{
+	serverError,
+	clientError,
+	success,
+}
+
+func resolveStatusCode(statusCode int, asyncOperation bool) requestStatus {
+	if asyncOperation {
+		if statusCode == 400 {
+			return serverError
+		}
+		return success
+	}
+	if statusCode > 500 && statusCode < 600 {
+		return serverError
+	}
+	if statusCode > 400 && statusCode < 500 {
+		return clientError
+	}
+	return success
+}
+
+func resolveEntityType(url url.URL) entity.Type {
+	entityType, _, _, _, _ := entity.ParseURL(url)
+	if entityType == entity.TypeContainer { // Handle containers as TypeInstance
+		return entity.TypeInstance
+	}
+	return entityType
+}
+
+type completedMetricsLabeling struct {
+	entityType entity.Type
+	status     requestStatus
+}
+
+var ongoingRequests map[entity.Type]*int64
+var completedRequests map[completedMetricsLabeling]*int64
+
+// InitAPIMetrics initializes maps with initial values for the API rates metrics.
+func InitAPIMetrics() {
+	ongoingRequests = make(map[entity.Type]*int64)
+	completedRequests = make(map[completedMetricsLabeling]*int64)
+
+	for _, entityType := range entity.EntityTypes {
+		if entityType != entity.TypeContainer { // We consider a container as an TypeInstance instead of TypeContainer
```

Comment:
We tend to use happy path coding, so rather than nesting everything here inside an if, instead use:

```go
if entityType == entity.TypeContainer {
    continue
}

// rest of code
```